### ISSUE TITLE
loki failing when querying huge data

### ISSUE THREAD

**Issue by isshwar** [grafana/loki#7084]
Hello,

loki querier is failing with the below when trying to pull logs (for example: last 30 days)

<html> <head><title>504 Gateway Time-out</title></head> <body> <center><h1>504 Gateway Time-out</h1></center> <hr><center>nginx/1.19.10</center> </body> </html>

We are running distributed set up with 4 querier pods each with 4Gi ram. when I try pull log data for 30days, seeing an error message on grafana. I have increased timeouts on nginx proxy that is between loki and grafana, still it didn't help.

On the other hand, when I try to pull for shorter periods and gradually increase it 30 days, able to see the logs. for example, I will start with 7 days, and then 12-14 days and then 20-25 days and in the end for 30 days. This way I could pull the logs, not sure if these are being cached in querier.

While our intention of testing loki is to replace ELK, but unless I would be able to atleast pull the data with timeouts, cannot completely switch from ELK. 

Do you have any ideas/suggestions on fixing this?

Thanks
Eswar

**Comment by yaroslavkasatikov** [comment]
Hi team, we have the same issue with loki and gw timeout.

We have our own-hosted s3 (ceph) and rather huge number of logs ( for example 40-50Gi per day).
We try to look 3h with easiest request (i.e {pod_labels_app=~"api-green|api-blue"} |= `site.com` ) and get timeout after the 3-4 mins.

`<html> <head><title>504 Gateway Time-out</title></head> <body> <center><h1>504 Gateway Time-out</h1></center> <hr><center>nginx</center> </body> </html>      
`

Try to understand what we do wrong..

our loki-distributed values file loks like:

```
# Values: https://github.com/grafana/helm-charts/tree/main/charts/loki-distributed
---
loki:
  structuredConfig:
    auth_enabled: false
    analytics:
      reporting_enabled: false
    chunk_store_config:
      max_look_back_period: 0s
    compactor:
      shared_store: filesystem
      # shared_store: aws
      working_directory: /var/loki/boltdb-shipper-compactor
    distributor:
      ring:
        kvstore:
          store: memberlist
    frontend:
      compress_responses: true
      log_queries_longer_than: 5s
      tail_proxy_url: http://loki-loki-distributed-querier:3100
    frontend_worker:
      frontend_address: loki-loki-distributed-query-frontend:9095
    ingester:
      chunk_block_size: 262144
      chunk_encoding: snappy
      chunk_idle_period: 30m
      chunk_retain_period: 1m
      lifecycler:
        ring:
          kvstore:
            store: memberlist
          replication_factor: 1
      max_transfer_retries: 0
      wal:
        dir: /var/loki/wal
    limits_config:
      enforce_metric_name: false
      max_cache_freshness_per_query: 10m
      reject_old_samples: true
      reject_old_samples_max_age: 168h
      split_queries_by_interval: 15m
    memberlist:
      join_members:
      - loki-loki-distributed-memberlist
    query_range:
      align_queries_with_step: true
      cache_results: true
      max_retries: 5
      results_cache:
        cache:
          enable_fifocache: true
          fifocache:
            max_size_items: 1024
            validity: 24h
    ruler:
      alertmanager_url: https://alertmanager.xx
      external_url: https://alertmanager.xx
      ring:
        kvstore:
          store: memberlist
      rule_path: /tmp/loki/scratch
      storage:
        local:
          directory: /etc/loki/rules
        type: local
    schema_config:
      configs:
      - from: "2022-07-11"
        store: boltdb-shipper
        object_store: s3
        schema: v11
        index:
          prefix: loki_index_
          period: 24h
        # object_store: filesystem
    server:
      http_listen_port: 3100
      grpc_server_max_recv_msg_size: 24194304
      grpc_server_max_send_msg_size: 24194304 

    storage_config:
      boltdb_shipper:
        active_index_directory: /var/loki/boltdb-shipper-active
        cache_location: /var/loki/boltdb-shipper-cache
        cache_ttl: 24h
        shared_store: s3
      aws:
        insecure: true
        s3: https://[path]/storage-prod
        s3forcepathstyle: true
        http_config:
          idle_conn_timeout: 90s
          response_header_timeout: 0s
          insecure_skip_verify: true
    
      # boltdb_shipper:
      #   active_index_directory: /var/loki/index
      #   cache_location: /var/loki/cache
      #   cache_ttl: 168h
      #   shared_store: filesystem
      # filesystem:
      #   directory: /var/loki/chunks
    table_manager:
      retention_deletes_enabled: false
      retention_period: 0s


queryFrontend:
  replicas: 3
compactor:
  persistence:
    enabled: true
    size: 100Gi
indexGateway:
  enabled: true
  persistence:
    enabled: true
    size: 100Gi
ingester:
  persistence:
    enabled: true
    size: 100Gi
querier:
  persistence:
    enabled: true
    size: 100Gi
ruler:
  persistence:
    enabled: true
    size: 100Gi
```

**Comment by isshwar** [comment]
thanks for your reply. 

I have only three labels in the my config 

`labels instance=$kubernetes['pod_name'],env=$env,app=$app`

My query looks like:

{env="prod",app="reading-metadata"} | json |= "log=error"

what I want to achieve?
while analysing issues, sometimes I might have to search with certain text(metadata related to the project) and may have to go back as long as 30 days.

**Comment by liguozhong** [comment]
Hi, I spent 1 year debugging the slow logql problem through tracing system. I have some experience you can refer to.

The performance of '| json' expr is too slow, you can pre-filter some logs by '|'

`{env="prod",app="reading-metadata"} |= "error"| json |= "log=error"
`
If the performance is still too slow, you can also speed up the query by replacing 'json' with 'regex'

`{env="prod",app="reading-metadata"} |= "error"| | regexp `"log":(?P<log>.*?)(}|,)` |= "log=error"
`

If it's still too slow, in addition to adding more machines, you can also try my branch, which is designed to solve the problem of grep being too slow
https://github.com/grafana/loki/pull/5455

**Comment by liguozhong** [comment]
`      split_queries_by_interval: 15m
`
Another supplementary point, if you do not have enough querier instances, or do not configure a large enough 'max_querier_per_tenant' value, this configuration will cause 7~30 days logql to become very slow.

**Comment by yaroslavkasatikov** [comment]
> Hi team, we have the same issue with loki and gw timeout.
> 
> We have our own-hosted s3 (ceph) and rather huge number of logs ( for example 40-50Gi per day). We try to look 3h with easiest request (i.e {pod_labels_app=~"api-green|api-blue"} |= 'site.com' ) and get timeout after the 3-4 mins.
> 
> `<html> <head><title>504 Gateway Time-out</title></head> <body> <center><h1>504 Gateway Time-out</h1></center> <hr><center>nginx</center> </body> </html> <!-- a padding to disable MSIE and Chrome friendly error page --> <!-- a padding to disable MSIE and Chrome friendly error page --> <!-- a padding to disable MSIE and Chrome friendly error page --> <!-- a padding to disable MSIE and Chrome friendly error page --> <!-- a padding to disable MSIE and Chrome friendly error page --> <!-- a padding to disable MSIE and Chrome friendly error page --> `
> 
> Try to understand what we do wrong..
> 
> our loki-distributed values file loks like:
> 
> ```
> # Values: https://github.com/grafana/helm-charts/tree/main/charts/loki-distributed
> ---
> loki:
>   structuredConfig:
>     auth_enabled: false
>     analytics:
>       reporting_enabled: false
>     chunk_store_config:
>       max_look_back_period: 0s
>     compactor:
>       shared_store: filesystem
>       # shared_store: aws
>       working_directory: /var/loki/boltdb-shipper-compactor
>     distributor:
>       ring:
>         kvstore:
>           store: memberlist
>     frontend:
>       compress_responses: true
>       log_queries_longer_than: 5s
>       tail_proxy_url: http://loki-loki-distributed-querier:3100
>     frontend_worker:
>       frontend_address: loki-loki-distributed-query-frontend:9095
>     ingester:
>       chunk_block_size: 262144
>       chunk_encoding: snappy
>       chunk_idle_period: 30m
>       chunk_retain_period: 1m
>       lifecycler:
>         ring:
>           kvstore:
>             store: memberlist
>           replication_factor: 1
>       max_transfer_retries: 0
>       wal:
>         dir: /var/loki/wal
>     limits_config:
>       enforce_metric_name: false
>       max_cache_freshness_per_query: 10m
>       reject_old_samples: true
>       reject_old_samples_max_age: 168h
>       split_queries_by_interval: 15m
>     memberlist:
>       join_members:
>       - loki-loki-distributed-memberlist
>     query_range:
>       align_queries_with_step: true
>       cache_results: true
>       max_retries: 5
>       results_cache:
>         cache:
>           enable_fifocache: true
>           fifocache:
>             max_size_items: 1024
>             validity: 24h
>     ruler:
>       alertmanager_url: https://alertmanager.xx
>       external_url: https://alertmanager.xx
>       ring:
>         kvstore:
>           store: memberlist
>       rule_path: /tmp/loki/scratch
>       storage:
>         local:
>           directory: /etc/loki/rules
>         type: local
>     schema_config:
>       configs:
>       - from: "2022-07-11"
>         store: boltdb-shipper
>         object_store: s3
>         schema: v11
>         index:
>           prefix: loki_index_
>           period: 24h
>         # object_store: filesystem
>     server:
>       http_listen_port: 3100
>       grpc_server_max_recv_msg_size: 24194304
>       grpc_server_max_send_msg_size: 24194304 
> 
>     storage_config:
>       boltdb_shipper:
>         active_index_directory: /var/loki/boltdb-shipper-active
>         cache_location: /var/loki/boltdb-shipper-cache
>         cache_ttl: 24h
>         shared_store: s3
>       aws:
>         insecure: true
>         s3: https://[path]/storage-prod
>         s3forcepathstyle: true
>         http_config:
>           idle_conn_timeout: 90s
>           response_header_timeout: 0s
>           insecure_skip_verify: true
>     
>       # boltdb_shipper:
>       #   active_index_directory: /var/loki/index
>       #   cache_location: /var/loki/cache
>       #   cache_ttl: 168h
>       #   shared_store: filesystem
>       # filesystem:
>       #   directory: /var/loki/chunks
>     table_manager:
>       retention_deletes_enabled: false
>       retention_period: 0s
> 
> 
> queryFrontend:
>   replicas: 3
> compactor:
>   persistence:
>     enabled: true
>     size: 100Gi
> indexGateway:
>   enabled: true
>   persistence:
>     enabled: true
>     size: 100Gi
> ingester:
>   persistence:
>     enabled: true
>     size: 100Gi
> querier:
>   persistence:
>     enabled: true
>     size: 100Gi
> ruler:
>   persistence:
>     enabled: true
>     size: 100Gi
> ```

@liguozhong  @jeschkies 
Team, could you please take a look into my config and give some advice? 
Don't want to create new issue, because the problem is the same -- too long query for s3 logs.  Scaled queriers up to 12 pods. They eat ~2Gb ram while search. Split them from index gateway. Added 3 query frontends.. No lucky.
The query for 3 hours got down with 504 Gateway timeout :(
Yes, we have rather many logs (~30G per day). Logs aren't json or something like this.. plain lines. sometimes - long lines. I have no ideas how to improve it. S3 based on Ceph with SSD, so rather fast disks. Haven't seen any issues with network (there are balancers pool before S3). No errors in log, except logs on frontend:

```
level=info ts=2022-09-07T13:39:50.815911581Z caller=handler.go:174 org_id=fake msg="slow query detected" method=GET host=loki-loki-distributed-query-frontend.services.svc.cluster.local:3100 path=/loki/api/v1/query_range time_taken=37.116422835s param_end=1662425999000000000 param_limit=1000 param_query="{pod_labels_app=~\"api-green|api-blue\"} |~ `.*OK`" param_start=1662411600000000000 param_step=10000ms param_direction=backward
```

**Comment by yaroslavkasatikov** [comment]
Want to add that if you search logs without filter, I mean query like `{pod_labels_app=~"api.*"}` - it works for 24h examples, but returned 504 for volumes (check screen)
<img width="640" alt="photo_2022-09-09_12-43-09" src="https://user-images.githubusercontent.com/7746585/189324984-b42f4372-5ec1-49d5-b558-584163c5ce7c.png">

But if I we add something like `{pod_labels_app=~"api.*"}|= "text"` - it leads to full timeout(

**Comment by isshwar** [comment]
@yaroslavkasatikov may i know if you running this kubernetes, could you let me know the replicas and resources you are provisioned for the stack.

@liguozhong still no luck. increased resources and still seeing the same error. while it is true that this could not be compared with ELK, but as we want to replace ELK with loki, atleast want to run the query to success. search criteria differs. unless, this doesn't throw 50* errors, cannot use this completely to replace ELK.

Also, for kubernetes distributed setup, do you have any suggestions on the replicas/resources?

below is the config loki config file. anything to improve. 

`
    auth_enabled: false
    compactor:
      shared_store: s3
      working_directory: /data/compactor
      compaction_interval: 10m
      retention_enabled: true
      retention_delete_delay: 2h
      retention_delete_worker_count: 150
      compactor_ring:
        kvstore:
          store: memberlist
    distributor:
      ring:
        kvstore:
          store: memberlist
    frontend:
      compress_responses: true
      log_queries_longer_than: 27s
    frontend_worker:
      frontend_address: loki-query-frontend:9095
      grpc_client_config:
        max_send_msg_size: 17179869184  # 16GiB
    querier:
     engine:
         timeout: 3m
         max_look_back_period: 40s
     query_timeout: 2m
     query_ingesters_within: 1s
    ingester:
      flush_check_period: 30s
      flush_op_timeout: 3m
      concurrent_flushes: 64
      chunk_block_size: 15728640
      chunk_encoding: lz4
      chunk_idle_period: 1m
      chunk_retain_period: 0m
      max_chunk_age: 5m
      lifecycler:
        ring:
          kvstore:
            store: memberlist
          replication_factor: 1
      max_transfer_retries: 0
      wal:
        dir: /data/loki/wal
    limits_config:
      max_entries_limit_per_query: 0
      ingestion_rate_strategy: global
      reject_old_samples: true
      ingestion_rate_mb: 100
      ingestion_burst_size_mb: 100
      max_cache_freshness_per_query: 1m
      creation_grace_period: 1m
      enforce_metric_name: false
      split_queries_by_interval: 15m
    memberlist:
      join_members:
        - loki-memberlist
    query_range:
      align_queries_with_step: true
      parallelise_shardable_queries: true
      cache_results: true
      max_retries: 5
      results_cache:
        cache:
          enable_fifocache: true
          fifocache:
            max_size_items: 1024
            validity: 24h
          memcached_client:
            host: ${PROJECT_NAME}-memcached-frontend.${ENVIRONMENT}.svc.k3s-test.local
            service: http
    ruler:
      alertmanager_url: ${ALERTMANAGER_URL}
      ring:
        kvstore:
          store: memberlist
      rule_path: /var/loki/rules-tmp
      storage:
        local:
          directory: /loki/rules
        type: local
      enable_api: true
      enable_alertmanager_v2: true
      enable_sharding: true

      alertmanager_client:
        basic_auth_username: "${ALERTMANAGER_USERNAME}"
        basic_auth_password: "$${q}{ALERTMANAGER_PASSWORD}"
    schema_config:
      configs:
        - index:
            period: 24h
            prefix: loki_index_
          object_store: s3
          schema: v11
          store: boltdb-shipper
    server:
      http_listen_port: 3100
      log_level: info
      grpc_server_max_recv_msg_size: 10485760000
      grpc_server_max_send_msg_size: 10485760000
      http_server_write_timeout: 10m
      http_server_read_timeout: 10m
      http_server_idle_timeout: 10m
      graceful_shutdown_timeout: 2m
    tracing:
      enabled: false
    storage_config:
      boltdb_shipper:
        active_index_directory: /data/loki/index
        cache_location: /data/loki/cache
        index_gateway_client:
          server_address: dns:///loki-index-gateway:9095
        shared_store: s3
      aws:
        access_key_id: ${S3_ACCESS_KEY_ID}
        bucketnames: ${S3_BUCKET_NAME}
        endpoint: https://pure-storage.pageplace.de
        s3forcepathstyle: true
        secret_access_key: $${q}{S3_SECRET_ACCESS_KEY}
        http_config:
          response_header_timeout: 5s
          insecure_skip_verify: true
      index_queries_cache_config:
        memcached:
          batch_size: 1000
          parallelism: 1000
        memcached_client:
          host: ${PROJECT_NAME}-memcached-index-queries.${ENVIRONMENT}.svc.k3s-test.local
          service: http
    chunk_store_config:
      chunk_cache_config:
          memcached:
            batch_size: 1000
            parallelism: 1000
          memcached_client:
            host: ${PROJECT_NAME}-memcached-chunks.${ENVIRONMENT}.svc.k3s-test.local
            service: http
      max_look_back_period: 0s
      write_dedupe_cache_config:
          memcached:
            batch_size: 1000
            parallelism: 1000
          memcached_client:
            host: ${PROJECT_NAME}-memcached-index-writes.${ENVIRONMENT}.svc.k3s-test.local
            service: http
`

**Comment by liguozhong** [comment]
![image](https://user-images.githubusercontent.com/9583245/189378038-538a2b86-4b96-479e-b4ee-d38e36c2e9ac.png)
![image](https://user-images.githubusercontent.com/9583245/189378386-240bceb4-b1eb-48c9-af56-fc1beaa3886a.png)


increase ‘max_query_parallelism’ and ‘max_concurrent’ add   ’split_queries_by_interval‘
split_queries_by_interval：15m  ，Your current configuration needs to split the query into a huge query load and parallelism.

 you need to show more error log. If error log is 504 timeout, it may also say the proxy timeout of grafana's loki datasource. By default, only 30s is a very low proxy time. This grafana datasource http proxy is used to forward logql queries.

**Comment by liguozhong** [comment]
I have successfully migrated our ELK to loki. It can completely replace the original function of ELK. Loki needs to be carefully tuned and expanded.

In the early stage, it took about 2 years to study the loki source code and deploy the loki cluster, and it took 9 months to migrate and offline ELK.

**Comment by liguozhong** [comment]
For the OLAP-type BI report of 30-day large query, it is recommended that you store the data to prometheus through loki's recording rule feature. loki is not very good at OLAP

**Comment by liguozhong** [comment]
If you need to pay attention to performance issues for a long time, I suggest you use the tracing system to trace the query of loki, and then cooperate with the source code of loki to see where it is slow.

**Comment by wilfriedroset** [comment]
Thank you for sharing your work around BI like workload on loki. I've a somehow similar usecase.
I would like to store an enriched [network flow](https://en.wikipedia.org/wiki/NetFlow) data model to provide a backend for network observability.
In the netflow we found, among other things:
* SrcAddr: 2^32 values (~4Billions)
* DstAddr: 2^32 values (~4Billions)
* SrcAS: 65K values
* DstAS: 65K values
* SrcPort: ~50K values (even if only few are wide used)
* DstPort: ~50K values (even if only few are wide used)

So even with only those 6 over the 50+ fields below, we have roughly a potential cardinality of ~2e+38.
We can assume that t he cardinality associated with netflow flow is infinite.

I’m wondering if loki is the best tool for that from the query perspective. The ingestion is doing really great.
However the query side, from grafana. I’m unable to execute large queries such as:
`rate({app="goflow2", way="egress"} | json | unwrap bytes | __error__="" [1s])` 
I’ve plenty of resources dedicated to each target, but they seems not used:
* 2 * query-scheduler 2 core/7GB
* 2 * query-frontend 2 core/7GB
* 20 * querier 16 core/60GB
* 4 * distributor 4 core/15GB
* 4 * ingester 16 core/60GB

I’ve been poking around the setting to get a working query, this is where I’m now.
Basically I’ve increase each limit which prevent the execution of the query above. I reckon this is not the best way to use efficiently the compute resource at my disposal.
```yaml
auth_enabled: False
server:
  grpc_listen_address: 10.0.3.41
  grpc_server_max_concurrent_streams: 1000
  grpc_server_max_recv_msg_size: 80000000
  grpc_server_max_send_msg_size: 80000000
  http_listen_address: 10.0.3.41
  http_listen_port: 3100
  http_server_idle_timeout: 40m
  http_server_read_timeout: 20m
  http_server_write_timeout: 20m
  log_format: json
  log_level: debug

distributor:
  ring:
    instance_interface_names:
    - ens4
    kvstore:
      prefix: loki/collectors/
      store: consul

querier:
  engine:
    timeout: 15m
  max_concurrent: 512
  query_timeout: 5m

ingester_client:
  grpc_client_config:
    grpc_compression: snappy
    max_recv_msg_size: 1048576000
    max_send_msg_size: 1048576000

ingester:
  lifecycler:
    interface_names:
    - ens4
    num_tokens: 128
    ring:
      kvstore:
        prefix: loki/collectors/
        store: consul
      replication_factor: 1
      zone_awareness_enabled: false

storage_config:
  boltdb:
    directory: /var/lib/loki/index
  filesystem:
    directory: /var/lib/loki/chunks

schema_config:
  configs:
  - from: 2020-05-15
    index:
      period: 168h
      prefix: index_
    object_store: filesystem
    schema: v11
    store: boltdb

limits_config:
  ingestion_burst_size_mb: 100
  ingestion_rate_mb: 50
  max_entries_limit_per_query: 20000
  max_queriers_per_tenant: 40
  max_query_parallelism: 5000
  max_query_series: 100000000
  per_stream_rate_limit: 30MB
  per_stream_rate_limit_burst: 60MB
  split_queries_by_interval: 5m

frontend_worker:
  grpc_client_config:
    grpc_compression: snappy
    max_recv_msg_size: 1048576000
    max_send_msg_size: 1048576000
  parallelism: 64
  scheduler_address: query-scheduler.loki-grpc.service.consul:9095

frontend:
  compress_responses: true
  grpc_client_config:
    grpc_compression: snappy
    max_recv_msg_size: 1048576000
    max_send_msg_size: 1048576000
  instance_interface_names:
  - ens4
  scheduler_address: query-scheduler.loki-grpc.service.consul:9095
  scheduler_worker_concurrency: 16
```
Neither my querier nor my query-frontend seems to be busy. Am I missing something obvious in my configuration?
The queriers seems to have a hard time to reach out to the scheduler as I have the following error message: `error notifying scheduler about finished query`
The queriers are sending huge reponse: `response larger than max message size
Then rpc error: code = ResourceExhausted desc = grpc: received message larger than max (100113955 vs. 80000000)`  despite the fact that the queriers managed to complete the query and I’ve this message `error notifying frontend about finished query`


For the context, I’m running loki 2.6.1 on baremetal (one target per host) with consul for the rings. The ingestion rate is stable at 220k/sec. Once I manage to solve the querying part I will back loki with an S3-like object storage.

**Comment by liguozhong** [comment]
If you are using ebpf to monitor network requests, I recommend pre-aggregating on the agent side, and then pushing the pre-aggregated metrics to prometheus. 
loki is only responsible for the search of detailed logs for‘ {app="bar"}|="404"’ .  metrics should be done by prometheus.

loki is not very very good at OLAP now

**Comment by liguozhong** [comment]
If you are using ebpf to monitor network requests, I recommend pre-aggregating on the agent side, and then pushing the pre-aggregated metrics to prometheus. loki is only responsible for the search of detailed logs for |="" . Metrics of type metrics should be done by prometheus....

old logql
`rate({app="goflow2", way="egress"} | json | unwrap bytes | __error__="" [1s])
`
new logql
```
sum(
       rate(
            {app="goflow2", way="egress"} 
                  | regexp `"bytes":(?P<bytes>.*?)(,|\{|\[)`
                  | unwrap bytes  
            [1m]) 
) by(app,way)
```

**Comment by wilfriedroset** [comment]
> If you are using ebpf to monitor network requests, I recommend pre-aggregating on the agent side, and then pushing the pre-aggregated metrics to prometheus.

I cannot pre-aggregate the data as my netops team queries highly depends on the situation at hand. As such prometheus is not suited for this due to the cardinality being almost unlimited.
We are investingating such tools:
* https://github.com/netobserv/flowlogs-pipeline
* https://github.com/netobserv/goflow2-loki-exporter
* https://github.com/netsampler/goflow2

In fact, I'm trying to use loki to mimic [kentik](https://www.kentik.com/) features.

In you rewritten query `new logql`, are you using regexp parser instead of the json due to the performance issue? If that is so, do you have experience with the logfmt parser from a performance perspective?

**Comment by liguozhong** [comment]
> > If you are using ebpf to monitor network requests, I recommend pre-aggregating on the agent side, and then pushing the pre-aggregated metrics to prometheus.
> 
> I cannot pre-aggregate the data as my netops team queries highly depends on the situation at hand. As such prometheus is not suited for this due to the cardinality being almost unlimited. We are investingating such tools:
> 
> * https://github.com/netobserv/flowlogs-pipeline
> * https://github.com/netobserv/goflow2-loki-exporter
> * https://github.com/netsampler/goflow2
> 
> In fact, I'm trying to use loki to mimic [kentik](https://www.kentik.com/) features.
> 
> In you rewritten query `new logql`, are you using regexp parser instead of the json due to the performance issue? If that is so, do you have experience with the logfmt parser from a performance perspective?

hi, You can pre-aggregate net data through promtail's 'metrics' pipeline, and only logql's 'metrics query' will not achieve the performance you expect.
https://grafana.com/docs/loki/latest/clients/promtail/stages/metrics/

**Comment by liguozhong** [comment]
> logfmt parser

'logfmt parser' has no performance issues as far as I know, but 'json parser' does have lower performance.

**Comment by isshwar** [comment]
@liguozhong 
Along with the applying suggestions that you have pointed out and increasing the system resoures, i am able to see gateway errors less times that earlier. for the time being, stopping it there as I  have been running in to another issue.

We have elk running parallel to loki. When I compare logs lines for the same query between loki and elk , I see loki is missing/not-reporting some log lines that I can on kibana. Wanted to know, if there is any issue with loki missing some log line to ingest to ingester or querier not returning all the loglines and some other config issue?

Thanks
Eswar

**Comment by liguozhong** [comment]
`sum by (tenant,reason) (rate(loki_discarded_bytes_total{}[1m]))>0`

It should trigger the ratelimit behavior of loki. You can check which reason through this promql, and then adjust the corresponding configuration threshold according to the reason.

more info： https://github.com/grafana/loki/pull/7145/files

![image](https://user-images.githubusercontent.com/9583245/191423806-2dfad47f-985b-4540-b2fe-d35b44639464.png)

**Comment by isshwar** [comment]
@liguozhong 

unfortunately, atleast in the last 24 hours, i don't see any discarded bytes after making the some changes that includes ´reject_old_samples: false´ and ´max_chunk_age: 120m´.
I have reduced the number of streams pausing the logs from prod cluster to rule out the fact that this is not due to clogging on the network. 
do you have any other suggestion?

**Comment by liguozhong** [comment]
The early loki version may have some inaccurate data bugs. What is your current loki version? 
In addition, I need to look at more detailed data (such as what logql is, and how long is the query time). 
I think it's hard to share such comprehensive data in this issue. This problem is difficult to solve in this issue.

**Comment by isshwar** [comment]
@liguozhong 

I am running 2.61. I didn't get you completely. Are you saying that the problem is difficult to solve? While looking to find more details on the issue, I found that there are other users who are struggling with the same issue. so, this is not a issue only with my config. Only, when I started to compare the log data between Elk and Loki, I got to know of this issue. 

Without fixing this, the migration would stall in my case. Do you think that this can be fixed?

**Comment by LinTechSo** [comment]
this issue I think is quite similar to the https://github.com/grafana/loki/issues/4015
I also used suggestions from that issue but nothing helped and the problem still persists.
and also, in my opinion, increasing the timeout configuration is not a solution.
Don't get to the core of the problem which is slow performance.

**Comment by LinTechSo** [comment]
Hi. 
I have an update about my test with the above traffic rate and below configuration. 
Any suggestion/advice would help.  thanks. 
my resources:
```
Querier:
    Memory:  6Gi
    CPU:      
          Request: 3
          Limit:    6
    Min Replica:    3
Ingester:
      Memory: 7Gi
      CPU:
            Request: 500m
            Limit:     1
      Min Replica:    2
Query FrontEnd:
        Memory: 1Gi
        CPU:
              Request: 1
              Limit: 2
        Min Replica: 2
Query Scheduler:
        Memory: 1Gi
         CPU:
                Request: 1
                Limit: 2
        Min Replica: 2
```
and my configuration

```
      max_recv/send_msg_size:     536870912000
      http_server_read/write/idle_timeout:     600
      proxy_read/send_timeout     300;
      proxy_connect_timeout     300;
      client_max_body_size     100M;
      Querier timeout:  10m
      max_query_parallelism:     32
      split_queries_by_interval:     15m
      Max_concurrent:     512
      Cache enabled
      Querier replica set to 3 with hpa
```
![Screenshot_2022-09-30_09-58-10](https://user-images.githubusercontent.com/53074289/193453911-e76a98bd-cb6e-461a-a6e7-e1c550bb551a.png)


![Screenshot_2022-09-30_09-51-05](https://user-images.githubusercontent.com/53074289/193453937-0f0f74c4-4dad-4dee-a12b-d6d4a47cbdf7.png)

appreciate it if you could help me figure this out.  how can I increase performance or at least get rid of 504 errors

**Comment by slim-bean** [comment]
The 504 error is coming from nginx, increasing timeouts in nginx should remove the 504 `proxy_read_timeout` iirc is the timeout to change

**Comment by slim-bean** [comment]
I just typed out a longer response but I'm on my phone and bumped a link and lost it 😬

The short version is that we are able to get query throughput out of Loki at rates of 50-100GB/s, we've had good success making Loki queries fast. Although there are many reasons from config to resources to storage that affect performance.

But not all storage is created equal... We use GCS on Google and S3 on Amazon which both work very well for highly parallelized workloads.

One of the examples in this thread is using a filesystem store which won't parallize at all, and I saw a mention of ceph which can be fast but we've also seen it slow, depends on how well it's built.

I'll try to follow up next week with some benchmark examples using the minio warp tool

**Comment by wilfriedroset** [comment]
> The short version is that we are able to get query throughput out of Loki at rates of 50-100GB/s, we've had good success making Loki queries fast. Although there are many reasons from config to resources to storage that affect performance.

I've been testing loki for a fair amount of time without great success. I cannot go above 1GB/s despite scaling the cluster vertically and horizontally. I've also poked around in the configuration to see what are the impact on the performance. I've also tried with and without memcached or redis which end in all ingesters going OOM.
My current conclusion is that tuning loki is not straighforward and I'm a bit lost. 
I understand it can be hard as you pointed not all storage are equals. On my end i'm using a S3 compatible, the same one I'm using with mimir. My mimir clusters are working like a charm.

@slim-bean would you be able to recommend a production ready configuration with sensible setting to get performance on read path and capacity planning, something similar to [mimir](https://grafana.com/docs/mimir/latest/operators-guide/run-production-environment/planning-capacity/)?

**Comment by Damien9527** [comment]
I meet the same issue. 

I have 1Tb log per day，  about 800 0000 count/ per min.   

My cluster info:
ec2 : 16c + 64Gb * 3 
storage: aws  s3 

ingest:  3
queirer: 3 
query-front: 3 
query schedule: 3 
gateway: 3 


when I query about 3 hour log ,  my ingest  get oom .   I don't know  how hug mem should I need . May someone share your cluster info and configuration.  Thanks.

**Comment by saibug** [comment]
It would be much appreciated if you put bench examples config

**Comment by StefanSa** [comment]
Hi there,
i had already pointed out this problem in May 2021, but no one could help here. https://github.com/grafana/loki/issues/3753
Also none of the core developers could help or provide a working configuration.
The discussion went on for more than 7 months, with the result that we stopped the migration to loki and now "successfully" switched to opensearch.

**Comment by Pajk** [comment]
Same issue for us. I'm testing Loki past few days (AWS EKS, S3 single store, official helm chart). I've tried numerous configurations but the aggregations are way too slow (sum by (x) (count_over_time()). I can't get over ~1.5GB throughput no matter how many resources or pods I use.

**Comment by suikast42** [comment]
I am using elk since 4 yeary for my bear metal setup and try to migrate to loki. 

I am facing the same query issue. But I  am using the monolithic mode with local file storage. I think that must fit the best performance for a single node setup ( I don't need realy a clustering at the moment).

That are the setting what makes loki useable for my usecase 
```yaml
querier:
  max_concurrent: 4096
query_scheduler:
  max_outstanding_requests_per_tenant: 10000

query_range:
  cache_results: true

limits_config:
  retention_period: 24h
```

My feeling about one month testing loki and reading docs is, that it can't realy replace the elk stack for my use case.  Let me tell why ( But I am fighting and hope to find a feasable setup to work with loki) :

1. The loki architecture is not easy to understand and to be honest not easy to scale or better to say plan the 'scale'.
2. You have to understand every single component in detail.   
3. Data retentions are hard to configure and uderstand. What is stored where ?  You have a cahce, boltdb a chunk a compactor that downloads uploads and deletes and so on. 
4. Capacitiy planing is more compilcated. For the compactor for example. 
5. You can't delete a single day ( somewhere middle in the month where a sensor produce a lot of waste for example) chunk out of box.
7. Maybe I miss something there. But why you guys are using an object storage for longterm store? I can't belive that can be performant for a high volume load like logging or metrics. ( The same for tempo and mimir ).  
8. I can't understand how loki guaranty data consistency in the cluster with the object storage.  I se some deduplication stuff in the querier but is that the way? Query everything what you have and drop the duplicates? 

I hope for someone who can teach me that I am doing and understanding everything wrong.

**Comment by fumanne** [comment]
I also  have this question... My  loki_distributor_bytes_received_total is about 10MB/s, when I query data within 6 hours, the query will be oom and my queriers replicas is 20， each limit cpu:2 memory:4Gi
--


<br class="Apple-interchange-newline">

**Comment by thejan2009** [comment]
Hey, I also dealt with this issue. Based on limited troubleshooting and going through some old docs, it seems changing the `limits_config.split_queries_by_interval` to `12h` or `24h` solves our problem.

This setting used to be part of [queryrange_config](https://grafana.com/docs/loki/v2.4.x/configuration/#queryrange_config) and had some conflicting comment above the default value:

```
# Split queries by an interval and execute in parallel, 0 disables it. You
# should use in multiple of 24 hours (same as the storage bucketing scheme),
# to avoid queriers downloading and processing the same chunks. This also
# determines how cache keys are chosen when result caching is enabled
# CLI flag: -querier.split-queries-by-interval
[split_queries_by_interval: <duration> | default = 30m]
```

Since changing the config value to `24h` solved the issue, I didn't dig any deeper to figure out if the claim in the comment is still valid or not.

**Comment by stecullum** [comment]
Does Grafana have a base-line environment (k8s, docker compose etc etc ) that can be used as a starting point eg..something with data generation, back end, loki configured etc etc 

At least we would have a baseline and some idea of the minimum throughput etc we can expect.

The perfomance issue seems to be a common one and we really have no baseline

**Comment by Daniel-Ebert** [comment]
> Hi, I spent 1 year debugging the slow logql problem through tracing system. I have some experience you can refer to.
> 
> The performance of '| json' expr is too slow, you can pre-filter some logs by '|'
> 
> `{env="prod",app="reading-metadata"} |= "error"| json |= "log=error" ` If the performance is still too slow, you can also speed up the query by replacing 'json' with 'regex'
> 
> `{env="prod",app="reading-metadata"} |= "error"| | regexp `"log":(?P.*?)(}|,)`|= "log=error"`
> 
> If it's still too slow, in addition to adding more machines, you can also try my branch, which is designed to solve the problem of grep being too slow #5455

Hi there, just to share my experience, this was very beneficial for me!

I changed the json-"Extraction" to Client Side via the Grafana transformations. This takes a huge load off from the central loki instance.
Probably, Regex would also be an option, I haven't tried out that so far.

**Comment by frittentheke** [comment]
> I changed the json-"Extraction" to Client Side via the Grafana transformations. This takes a huge load off from the central loki instance. Probably, Regex would also be an option, I haven't tried out that so far.

Could you maybe run Promtail with `-server.profiling_enabled` and, as described in https://github.com/grafana/loki/issues/8054, access http://localhost:9080/debug/pprof/heap to see which function does use up all those CPU cycles?

For an example of how this can help to find a regex performance issue see https://github.com/kiali/kiali/issues/4592

**Comment by Daniel-Ebert** [comment]
> > I changed the json-"Extraction" to Client Side via the Grafana transformations. This takes a huge load off from the central loki instance. Probably, Regex would also be an option, I haven't tried out that so far.
> 
> @Daniel-Ebert Could you maybe run Promtail with `-server.profiling_enabled` and, as described in #8054, access http://localhost:9080/debug/pprof to see which function does use up all those CPU cycles?
> 
> For an example of how this can help to find a regex performance issue see [kiali/kiali#4592](https://github.com/kiali/kiali/issues/4592)

Hi there @frittentheke,

sorry if there might have been some kind of misunderstanding, I was not referring to the promtail Performance, but rather to the "Query-Time LogQL Regex Parser" that goes to the Loki Backend.

Best regards

Daniel

**Comment by frittentheke** [comment]
> sorry if there might have been some kind of misunderstanding, I was not referring to the promtail Performance, but rather to the "Query-Time LogQL Regex Parser" that goes to the Loki Backend.

Sorry, I meant to say Loki.
Grafana themselves found quite a few performance issues using pprof - https://grafana.com/blog/2019/08/15/lokis-path-to-ga-query-optimization-part-one/


 Just turn it on and see what you can find.

**Comment by LinTechSo** [comment]
Hi all,
i want to ask to share your experiences with recent changes and improvements in Loki versions 2.8 and 2.9, regarding to this issue. Is this issue still open?
thanks

**Comment by illthizam-healthhelper** [comment]
same issue 


logger=ngalert.scheduler rule_uid=d4ea7eaa-caea-43b6-a7dd-2515f1ea1578 org_id=1 version=9 fingerprint=cbfe9c215a3b3871 attempt=0 now=2024-06-20T06:00:00Z rule_uid=d4ea7eaa-caea-43b6-a7dd-2515f1ea1578 org_id=1 t=2024-06-20T06:00:30.157796042Z level=error msg="Failed to evaluate rule" error="[sse.dataQueryError] failed to execute query [A]: Get \"http://hh-staging-grafana-loki-gateway.hh-staging-monitoring.svc.cluster.local:80/loki/api/v1/query_range?direction=backward&end=1718863200000000000&query=sum+by%28app%2C+pod%2CerrorMessage%29+%28count_over_time%28%7Bnamespace%3D%22hh-staging%22%7D+%7C%3D+%60%22statusCode%22%3A500%60+%7C+json+%7C+line_format+%60%60+%7C+__error__+%21%3D+%60JSONParserErr%60+%5B2m%5D%29%29+%3E++0&start=1718862600000000000&step=1000ms\": context deadline exceeded" duration=30.130208375s

**Comment by someStrangerFromTheAbyss** [comment]
Ok so im gonna share some of my recent experience with Loki and debugging a similar issue. In my case, clients were getting timeouts whenever they were doing 6hours + query. Im gonna put some of our mitigation, but i warn you this is still on going. Our stack has been designed for max 2 week retention so we are doing everything around that. First, here is our setup:

- Grafana loki 3.2.1
- Deployed using helm, using the chart version 6.10.0
- The deployment type is Simple Scalable
- Our cloud provider is an internal one using Openstack
- Our current storage is Azure blob storage. It was something else before which i will come back later.
- We have automatic scaling on the writes using CPU/RAM and the read automatically scale using some custom metrics recommended by the Grafana loki team in there documentation.

With that out of the way, here is some things we did to improve things:

- We changed our storage from our internal S3 compatible API to Azure Blob storage. Behind that S3 compatible API was Swift storage. Although our internal Storage had better response time since it was closer to our actual infra, we had big problems debugging when things were going bad. We don't have access to the S3 API logs so when we were getting timeouts, it was to know who was at fault. We decided to move to Azure Storage and since then, we have less errors for everything that is storage related. I also found out way later that grafana loki kinda not recommend S3 compatible API, you can see [here](https://grafana.com/docs/loki/latest/operations/storage/#-supported-chunks-stores-not-typically-recommended-for-production-use). So i would use extreme caution when using S3 compatible storage unless you really know what you are doing.
- Too many small chunks can impact query performance by a lot. I had client remark that his stack was having a long time to load labels and/or gave him timeouts/error 500. Turns out, we had the `shard_streams` function enabled with a very low `desired_rate`. This created lots of small chunks that made loki have a hard time to query the logs/labels. We since then increased the `desired_rate` to a higher threshold since then and are thinking on increasing `chunk_target_size` if we see that the clients is having way higher traffic and starts creating a lot of chunks.
- Ressources attributed to your loki can also impact query performance immensely. In our case, we were scaling the write pods on incoming bytes at first. This made our stack very lean, but, when clients were making lots of big queries, we were getting timeouts, write pods crashing,etc. So scaling the write based on CPU/RAM removed that problem. The write pods can be impacted by query more then you think.
- Of course, talking about ressources, having `limits` on your loki pods is a dangerous thing. We had some at the beginning that were matching the `requests` since loki can be quite ressources intensive from time to time, but we had to climb those up to 3Core since the spike in ressource can be so sudden that our grafana dashboard could not even show the CPU spike. The write pods were crashing or stopped responding even before we could see how much ressources it needed. So Kubernetes limit is something to keep in mind.
- Has for loki configuration, some of the loki configuration lacks details on what it does or what value you should put. In perticular, the `split_queries_by_interval` and `querier.max-concurrent`. The `split_queries_by_interval` needs to be changed depending on your query range you aim. We are aiming for short term range so1h was enough (before it was at 0). If you aim for 30days +, maybe changing the number to 24h will help query performance especially when it comes for metrics on long time ranges. `querier.max-concurrent` Should be the double of the available Core you have. So if you have 3Core CPU, it should be 6, if 2, it should be 4,etc. On that front, especially if you are coming from Elasticsearch, following this [guide](https://medium.com/engenharia-arquivei/grafana-loki-our-journey-on-replacing-elastic-search-and-adopting-a-new-logging-solution-at-f65aec407e47) more closely might help you get similar results.
- Finally, increasing timeouts. `http_server_read_timeout`, `http_server_write_timeout`, `http_server_idle_timeout `should be increased.

On a personnal note, coming from Elasticsearch/Opensearch, Loki is definetely better when it comes to ressources management, scaling and make it work fast. But, when it comes to querying, you have to do a lot of thinkering esepcially if you are trying to simulate some of the behavior of ES. Lots of changes on the configuration might be necessary to get the results for querying.

Hope it helps some people on there issue.

**Comment by someStrangerFromTheAbyss** [comment]
After some additionnal digging, i might have found the source of the problem, but it could be another issue. Please, feel free to link me to the correct issue to maybe help the people that have that problem.

So i found out that even on small data sample, the query can time out or finish on a 504 gateway timeout. 

First, we found something perticular, an http code 499 that appears on our Nginx when we hit that problem:
![Image](https://github.com/user-attachments/assets/114c9e37-5137-4b31-88cc-55c8fd2963d3)

Second, queries performance and timeouts appears right after an helm upgrade/deploy. We are using ansible to push our helm charts, so the module `kubernetes.core.helm`

Third, this was the error log when doing the query:
level=error ts=2024-11-05T19:19:02.128307681Z caller=retry.go:107 org_id=fake msg="error processing request" try=1 type=logproto.IndexStatsRequest query="CONFIDENTIAL" query_hash=1408452637 start=2024-11-05T19:03:31Z end=2024-11-05T19:19:01Z start_delta=15m31.128261424s end_delta=1.128261701s length=15m30s retry_in=1.288553305s code=Code(500) err="rpc error: code = Code(500) desc = rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing: dial tcp: lookup loki-backend-headless.6723a512e7641cd9c37269ed.svc.cluster.local on 172.20.0.10:53: no such host\""

**How to solve the issue:**

Simply restarting the read pods removed the 499 http code and made query performance back to normal.

Problem is, i still dont know how to prevent that issue. It is widly inconsistent, sometimes the read pods are fine after an helm change, sometimes they are not.  Maybe that explains why some people have this problem, while others do not.

If this can help, here is my loki config after deployment:
```
auth_enabled: false
common:
  compactor_address: 'http://loki-backend:3100'
  path_prefix: /var/loki
  replication_factor: 1
  storage:
    azure:
      account_key: PASSWORD
      account_name: ACCOUNT
      container_name: loki
      request_timeout: 30s
      use_federated_token: false
      use_managed_identity: false
compactor:
  compaction_interval: 10m
  delete_request_store: azure
  retention_delete_delay: 2h
  retention_delete_worker_count: 150
  retention_enabled: true
  working_directory: /tmp
frontend:
  scheduler_address: ""
  tail_proxy_url: ""
frontend_worker:
  scheduler_address: ""
index_gateway:
  mode: simple
ingester:
  chunk_idle_period: 30m
  chunk_target_size: 1572864
  flush_check_period: 15s
  wal:
    replay_memory_ceiling: 1024MB
limits_config:
  allow_structured_metadata: true
  ingestion_burst_size_mb: 15
  ingestion_rate_mb: 30
  max_cache_freshness_per_query: 10m
  max_chunks_per_query: 100
  max_concurrent_tail_requests: 100
  max_entries_limit_per_query: 1000
  max_global_streams_per_user: 50000
  max_label_names_per_series: 17
  max_query_parallelism: 128
  max_query_series: 50
  max_streams_matchers_per_query: 100
  per_stream_rate_limit: 5Mb
  per_stream_rate_limit_burst: 20Mb
  query_timeout: 300s
  reject_old_samples: true
  reject_old_samples_max_age: 168h
  retention_period: 14d
  shard_streams:
    desired_rate: 2000005
    enabled: true
  split_queries_by_interval: 1h
  tsdb_max_bytes_per_shard: 2GB
  tsdb_max_query_parallelism: 2048
  volume_enabled: true
memberlist:
  join_members:
  - loki-memberlist
pattern_ingester:
  enabled: false
query_range:
  align_queries_with_step: true
ruler:
  alertmanager_url: URL
  enable_alertmanager_v2: true
  enable_api: true
  storage:
    azure:
      account_key: PASSWORD
      account_name: ACCOUNT
      container_name: loki
      request_timeout: 30s
      use_federated_token: false
      use_managed_identity: false
    type: azure
runtime_config:
  file: /etc/loki/runtime-config/runtime-config.yaml
schema_config:
  configs:
  - from: "2024-07-29"
    index:
      period: 24h
      prefix: index_
    object_store: azure
    schema: v13
    store: tsdb
server:
  grpc_listen_port: 9095
  grpc_server_max_recv_msg_size: 60000000
  grpc_server_max_send_msg_size: 60000000
  http_listen_port: 3100
  http_server_idle_timeout: 600s
  http_server_read_timeout: 600s
  http_server_write_timeout: 600s
  log_level: debug
storage_config:
  hedging:
    at: 250ms
    max_per_second: 20
    up_to: 3
  tsdb_shipper:
    index_gateway_client:
      server_address: loki-backend-headless.loki.svc.cluster.local:9095
tracing:
  enabled: false
```
Any feeback or comments on this issue would be appreciated.

**Comment by someStrangerFromTheAbyss** [comment]
So i think im finally done with this issue. So after a lot of debugging, i concluded the source of the problem is either:

- Having both a boltdb-shipper and tsdb indexed setup
- Having a boltdb-shipper indexed set in your config even though you don't use it anymore

Lots of the timeout/499 seems to come from the fact that the read and the backend pods have some difficulty at startup to establish a connection. I blame the boltdb-shipper since having this setup for your backend makes the backend pod take up to 5 minutes to shutdown  and takes almost 1 minutes to be fully ready. I assume the first read pod makes a bad connection to the backend pod before actually ready for its boltdb-shipper part. But, thats just a theory since to prove it, i would actually setup something to intercept requests between pods and since for me, the problem is solved, ill stop here.

So basically, i just removed everything boltdb-shipper from my config. So i went from:
```
schema_config:
  configs:
  - from: "2022-01-11"
    index:
      period: 24h
      prefix: loki_index_
    object_store: azure
    schema: v12
    store: boltdb-shipper
  - from: "2024-07-29"
    index:
      period: 24h
      prefix: index_
    object_store: azure
    schema: v13
    store: tsdb
```
To:
```
schema_config:
  configs:
  - from: "2024-07-29"
    index:
      period: 24h
      prefix: index_
    object_store: azure
    schema: v13
    store: tsdb
```

After that, the other problem was the helm chart. Seem that even though you have no more boltdb-shipper configured, the helm chart by default tries to establish a boltdb-shipper connection to the backend pods even though its not set. To remedy this, i simply did this on my helm chart value file:
```
loki:
....
  storage_config:
    boltdb_shipper: null
....
```
To make sure that there is not boltdb-shipper connection made to the backend.

With that done, it seems for me at least, the query problem is gone. I havent tried again query large amounts of data like before, but for small scale problem query, the fix works. The fix should be pushed to clients next week and hopefully no new report comes in for large query problems.

Maybe my issue was not put in the correct issue thread, since there is a lot of other issue similar like this one for query timeouts, but since this one is like the first one that pops up on google, i assume its the most popular and that my conclusion might help someone with a similar issue. Feel free to redirect me to post my solution for my timeout issue to the correct one.

Hopefully it helps solved the other one this thread also.

**Comment by abhinavthula** [comment]
Hi, we are new to Loki,
We have Grafana deployed through Helm and deployed Loki as with Operator in the ARO cluster.

We are facing the similar issue, i.e., 

![Image](https://github.com/user-attachments/assets/5f31a080-f11d-4f6d-8e50-233219d48319)

Note: I tried increasing the timeout settings for Grafana however, it didn't resolve our issue

Adding the loki stack yaml below.


apiVersion: loki.grafana.com/v1
kind: LokiStack
metadata:
  annotations:
   
  managedFields:
    - apiVersion: loki.grafana.com/v1
      fieldsType: FieldsV1
      fieldsV1:
        'f:spec':
          .: {}
          'f:limits':
            .: {}
            'f:global':
              .: {}
              'f:ingestion':
                .: {}
                'f:ingestionBurstSize': {}
                'f:ingestionRate': {}
                'f:perStreamRateLimit': {}
                'f:perStreamRateLimitBurst': {}
              'f:queries':
                .: {}
                'f:queryTimeout': {}
          'f:managementState': {}
          'f:size': {}
          'f:storage':
            .: {}
            'f:schemas': {}
            'f:secret':
              .: {}
              'f:name': {}
              'f:type': {}
          'f:storageClassName': {}
          'f:tenants':
            .: {}
            'f:mode': {}
            'f:openshift':
              .: {}
              'f:adminGroups': {}
      manager: Mozilla
      operation: Update
      time: '2024-11-14T06:34:47Z'
    - apiVersion: loki.grafana.com/v1
      fieldsType: FieldsV1
      fieldsV1:
        'f:metadata':
          'f:annotations':
            .: {}
            'f:loki.grafana.com/certRotationRequiredAt': {}
            'f:loki.grafana.com/rulesDiscoveredAt': {}
      manager: manager
      operation: Update
      time: '2024-11-26T10:04:05Z'
    - apiVersion: loki.grafana.com/v1
      fieldsType: FieldsV1
      fieldsV1:
        'f:status':
          .: {}
          'f:components':
            .: {}
            'f:compactor':
              .: {}
              'f:Ready': {}
            'f:distributor':
              .: {}
              'f:Ready': {}
            'f:gateway':
              .: {}
              'f:Ready': {}
            'f:indexGateway':
              .: {}
              'f:Ready': {}
            'f:ingester':
              .: {}
              'f:Ready': {}
            'f:querier':
              .: {}
              'f:Ready': {}
            'f:queryFrontend':
              .: {}
              'f:Ready': {}
          'f:conditions': {}
          'f:storage':
            .: {}
            'f:credentialMode': {}
            'f:schemas': {}
      manager: manager
      operation: Update
      subresource: status
      time: '2024-11-26T10:04:05Z'
  name: logging-loki
  namespace: openshift-logging
  resourceVersion: '157849118'
  uid: af894837-ff43-449b-93e1-e6bd0e28679d
spec:
  limits:
    global:
      ingestion:
        ingestionBurstSize: 32
        ingestionRate: 16
        perStreamRateLimit: 5
        perStreamRateLimitBurst: 20
      queries:
        queryTimeout: 5m
  managementState: Managed
  size: 1x.extra-small
  storage:
    schemas:
      - effectiveDate: '2024-07-23'
        version: v13
    secret:
      name: logging-loki-azure
      type: azure
  storageClassName: managed-disk-premium
  tenants:
    mode: openshift-logging
    openshift:
      adminGroups:
        - cluster-admin
status:
  components:
    compactor:
      Ready:
        - logging-loki-compactor-0
    distributor:
      Ready:
        - logging-loki-distributor-69dd9b8bdf-r55st
        - logging-loki-distributor-69dd9b8bdf-xcp52
    gateway:
      Ready:
        - logging-loki-gateway-787588d979-5rw2h
        - logging-loki-gateway-787588d979-874p4
    indexGateway:
      Ready:
        - logging-loki-index-gateway-1
        - logging-loki-index-gateway-0
    ingester:
      Ready:
        - logging-loki-ingester-1
        - logging-loki-ingester-0
    querier:
      Ready:
        - logging-loki-querier-6cc7975d76-x9b8l
        - logging-loki-querier-6cc7975d76-6f8md
    queryFrontend:
      Ready:
        - logging-loki-query-frontend-bc8d6db5b-dc7rh
        - logging-loki-query-frontend-bc8d6db5b-k6h8r
  conditions:
    - lastTransitionTime: '2024-11-26T10:04:05Z'
      message: 'All components are running, but some readiness checks are failing'
      reason: PendingComponents
      status: 'False'
      type: Pending
    - lastTransitionTime: '2024-11-26T10:04:05Z'
      message: One or more LokiStack components failed
      reason: FailedComponents
      status: 'False'
      type: Failed
    - lastTransitionTime: '2024-11-26T10:04:05Z'
      message: All components ready
      reason: ReadyComponents
      status: 'True'
      type: Ready
  storage:
    credentialMode: static
    schemas:
      - effectiveDate: '2024-07-23'
        version: v13


Kindly let me know if something is missing here. Thank you

**Comment by abhinavthula** [comment]
Dear All,

Any inputs from anyone?

**Comment by WesselAtWork** [comment]
Getting same issue as @abhinavthula 

I think it might be the default query timeout 
https://github.com/grafana/loki/blob/9f25be15bdbe20bfd383a01add9982a56aa66d84/production/helm/loki/values.yaml#L348

300s sounds like a lot.
changed to `60s` to align with my grafana loki client settings (`1m`)

I also changed 
https://github.com/grafana/loki/blob/9f25be15bdbe20bfd383a01add9982a56aa66d84/production/helm/loki/values.yaml#L340-L341

To a (in my opinion) more reasonable `60s`

I think the problem here is "inflight" queries. I think the querier has a max number of queries that i can handle concurrently, the very big timeout means that there are a lot of queries "inflight" at any one momment.

I still get a lot of "context cancelled" errors in my logs for queries, but now the 503 has gone away.

**Comment by WesselAtWork** [comment]
## Update: 
it's still happening for me (but with less frequency, the big query limits seem to have aggravated the real issue).

## Experimentation
I can cause the issue to occur when I search a period of time with "fat" logs, i.e. logs that contain more text data then usual.
I then gradually reduced the line limit down during that period until the call succeeded, that is when I noticed that, the download logs were almost exactly 8Mi. (i.e. the results size was just under my `grpc_server_max_recv_msg_size`)

In my logs there were also a bunch of ResourceExhausted (see #2271)
> Weird that it is exceeding at all, isn't the call [streamed](https://jbrandhorst.com/post/grpc-binary-blob-stream/)?

That means that if you have a period of time for a given query where 1000 logs will exceed the GRPC message size, you will experience this issue.

Enabling trace and investigating further: I noticed something weird: _The query seemed to be split, but it was always executing on one reader._

---

I looked through the config parameters and found `tsdb_max_bytes_per_shard` which has a default of 600MB.
I then set this to `7MB` and, while the queries DID get properly spread now, it didn't fix the underlining issue.

I also tried `max_querier_bytes_read: 6MB` but that didn't seem to do anything. 
> `max_query_bytes_read` did what the description said. (It stops any query from running over timespans with the specified size)

the only recourse was to increase `grpc_server_max_recv_msg_size` which "solves" the problem, but I don't believe this is a solution.

**Comment by someStrangerFromTheAbyss** [comment]
We had this issue and we just resolved it. We can now query text on 100GB queries without too much problems. It does take a while though (Like 3 minutes) but it does work. Debugging this issue is a mess and depends on your infra and i can understand why Loki dev might relunctant to address this github issue since the root cause of the issue might be outside of the Loki. Here is our steps that we applied after multiple weeks of debugging.

Context: We deploy Loki in Simple Scalable mode in Openstack with Cilium networking. We have a Grafana in another Kubernetes cluster that query said Loki. We aim for 5 minutes Max Queries

**### Step 1: increase timeouts and msg size on Loki**

Most people already mentionned it, but i feel repeating it here might be worth it. Here is all the timeout/msg size you need to change on Loki:
```
server:
  grpc_server_max_recv_msg_size: DEPENDS_ON_INFRA
  grpc_server_max_send_msg_size: DEPENDS_ON_INFRA
  http_server_idle_timeout: 300s
  http_server_read_timeout: 300s
  http_server_write_timeout: 300s
limits_config:
  query_timeout: 300s
common:
  storage:
    STORAGE_TYPE:
      request_timeout: 300s
```

The grpc_server_max_recv_msg_size and grpc_server_max_send_msg_size depends on your infra and max message size on your nginx. We put 80000000 and 60000000 respectfully, but it will depend also on the number of log lines you expect to return to grafana. On our side its set at 1000, but if you put 10 000, you might need to increase this number. Cilium or whatever networking solution you use for your infra might also dictate the size.

**### Step 2: increase timeouts on the nginx and loadbalancer in front of the loki**

I insist on the load balancer. Even if you are in another cloud provider be Azure or AWS, **double check the timeouts**. In OpenStack, we set the following:

- Client Data Timeout: 300000
- Member Data Timeout: 300000

Second, the nginx. You probably have an nginx and for that, make sure you have the following annotations or a corresponding nginx conf:

```
    nginx.ingress.kubernetes.io/proxy-connect-timeout: '300'
    nginx.ingress.kubernetes.io/proxy-next-upstream-timeout: '300'
    nginx.ingress.kubernetes.io/proxy-read-timeout: '300'
    nginx.ingress.kubernetes.io/proxy-send-timeout: '300'
```

You can now test Loki directly **whitout going through grafana**. You should be able to do 5 minutes query directly to Loki at this point. If it passes, you can now go to step 3.

**### Step 3: increase timeouts on the nginx and loadbalancer in front of the Grafana. Increase Grafana timeouts**

You will then need to do the same thing you did for the load balancer/nginx in front of the Loki but now, on Grafana's side. The nginx and the Loadbalancer needs increased timeout or else, you might get Timeout/Gateway problems once more.

Then, either you change the .ini file of the Grafana to have:

```
[dataproxy]

timeout = 300
```

Or, change the timeout directly in the datasource.

Once that is done, you should finally be able to query big data/do huge queries that gets no timeout.

**### Step 4, things to look out for**

- Make sure all the Read pods are functionning: Just had a MR closed that added StartupProbe for the read pods since it seems there was a problem with read pods at startup during dpeloyment. Maybe thats your case. Check the [issue](https://github.com/grafana/loki/issues/15191).
- If what you are trying to do is to migrate from ELK to Loki, i would recommend to find a guide on that specifically. From what i have read on the subject, you might need to deploy in microservice mode, have a memory cache and have a an approach of one big Loki with multi tenant enabled.

**Comment by Spinestars** [comment]
1. I found that if the length of a log is extremely long, say more than 64KB, the query will get stuck on this log. The default chunk size for Loki is 256KB.
2. Enable more chunksCache 
I am looking for a way that, when using promtail or alloy, and encountering extremely long single-line logs, promtail or alloy can split them and convert them into multiple lines.
